source/h2_stream.c

```diff
@@ -871,6 +880,38 @@ struct aws_h2err aws_h2_stream_on_decoder_data_begin(
         return s_send_rst_and_close_stream(stream, aws_h2err_from_h2_code(AWS_HTTP2_ERR_PROTOCOL_ERROR));
     }
 
+    if (stream->thread_data.content_length_received) {
+        stream->thread_data.received_data_length += payload_len;
+        /**
+         * RFC-9113 8.1.1:
+         * A request or response is also malformed if the value of a content-length header field does not equal the
+         * sum of the DATA frame payload lengths that form the content, unless the message is defined as having no
+         * content.
+         *
+         * Clients MUST NOT accept a malformed response.
+         */
+        if (stream->thread_data.received_data_length > stream->thread_data.content_length) {
+            AWS_H2_STREAM_LOGF(
+                ERROR,
+                stream,
+                "Total received data payload=%" PRIu64 " has exceed the received content-length header, which=%" PRIi64
+                ". Closing malformed stream",
+                stream->thread_data.received_data_length,
+                stream->thread_data.content_length);
+            return s_send_rst_and_close_stream(stream, aws_h2err_from_h2_code(AWS_HTTP2_ERR_PROTOCOL_ERROR));
+        }
+        if (end_stream && stream->thread_data.received_data_length != stream->thread_data.content_length) {
+            AWS_H2_STREAM_LOGF(
+                ERROR,
+                stream,
+                "Total received data payload=%" PRIu64
+                " does not match the received content-length header, which=%" PRIi64 ". Closing malformed stream",
+                stream->thread_data.received_data_length,
+                stream->thread_data.content_length);
+            return s_send_rst_and_close_stream(stream, aws_h2err_from_h2_code(AWS_HTTP2_ERR_PROTOCOL_ERROR));
+        }
```

Comment:
we might need to do this check elsewhere
it's possible for on_data_begin to never fire
the message might declare content-length and then send END_STREAM in the HEADERS frame
maybe do the `!=` check in the end_stream callback?

also, check out this language:
> unless the message is defined as having no content.

which makes me think of a HEAD request where the response has content-length but no actual data :|
check how we do it in HTTP/1